File: src/main/java/enju/ccg/ml/Perceptron.java
package enju.ccg.ml;

import java.util.List;
import enju.ccg.util.Util;
import enju.ccg.util.Pair;

public class Perceptron<L> extends AbstractClassifier<L> {
  private WeightVector averageWeight;
  private double C; // count for averaging
  public Perceptron(WeightVector weight) {
    super(weight);    
    averageWeight = new WeightVector();
    C = 1;
  }

  // multiclass classification
  public void update(Example<L>[] examples, L gold) {
    L pred = predict(examples).getP1();
    if (pred != gold) {
      updateBody(examples, pred, gold);
    }
  }
  // multiclass classification
  public void updateBody(Example<L>[] examples, L pred, L gold) {
    for (Example<L> e : examples) {
      if (e.getLabel().equals(pred)) {
        for (int f : e.getFeature()) {
          weight.add(f, -1);
          averageWeight.add(f, -C);
        }
      } else if (e.getLabel().equals(gold)) {
        for (int f : e.getFeature()) {
          weight.add(f, 1);
          averageWeight.add(f, C);
        }
      }
    }
    C += 1;
  }
  // structured perceptron
  public void update(int[] predFeatures, int[] goldFeatures) {
    for (int f : predFeatures) {
      weight.add(f, -1);
      averageWeight.add(f, -C);
    }
    for (int f : goldFeatures) {
      weight.add(f, 1);
      averageWeight.add(f, C);
    }
    C += 1;
  }
  public void finalize() {
    System.out.println(C);
    for (int i = 0; i < weight.size(); ++i) {
      weight.set(i, weight.get(i) - averageWeight.get(i) / C);
    }
  }
}

File: src/main/scala/enju/ccg/ml/Perceptron.scala
package enju.ccg.ml

class Perceptron[L](override val weights: NumericBuffer[Float]) extends LinearClassifier[L] with OnlineTrainer[L] {
  val averageWeights = new NumericBuffer[Float](weights.size)
  var c = 1.0F

  override def update(examples: Seq[Example[L]], gold: L): Unit = {
    val pred = predict(examples)._1
    if (pred != gold) {
      var i = 0
      while (i < examples.size) {
        val label = examples(i).label
        if (label == pred) updateFeatureWeighs(examples(i).featVec, -1.0F)
        else if (label == gold) updateFeatureWeighs(examples(i).featVec, 1.0F)
        i += 1
      }
    }
    c += 1.0F
  }
  def updateFeatureWeighs(featVec: Array[Int], scale: Float): Unit = featVec.foreach { f =>
    weights(f) += scale
    averageWeights(f) += scale * c
  }
  def update(predFeatVec:Array[Int], goldFeatVec:Array[Int]): Unit = {
    updateFeatureWeighs(predFeatVec, -1.0F)
    updateFeatureWeighs(goldFeatVec, 1.0F)
    c += 1.0F
  }
  def takeAverage: Unit = (0 until weights.size) foreach { i =>
    weights(i) -= averageWeights(i) / c
  }
}
